Simplify zero-shot topic modeling

ianrandman opened the pull request:
* Fix downstream operations after zero-shot topic modeling
  * No longer perform merging of models for zero-shot. The zero-shot topics act as a prefiltering step before clustering. The zero-shot topics are combined with the clustered topics immediately after clustering.
  * The C-TF-IDF model is fitted on all documents now, regardless of whether they belong to clustered or zero-shot topics.
* Validate number of topics when using zero-shot topic modeling
* Remove check for `type(self.hdbscan_model) != BaseCluster` when checking whether model is zero-shot
  * The HDBSCAN model is replaced with a `BaseCluster` during `fit_transform()` during zero-shot topic modeling.
* Derive `self._outliers` rather than tracking it to maintain alignment using `@property`
* Derive zero-shot labels when requested rather than tracking it using `@property`
* Fix typos related to `topic_to`, `topics_from` for mapping
* Validate existence of outliers in `reduce_outliers()`
* Maintain zero-shot topics while reducing topics
  * If original topics contain one or more zero-shot topics, the new topic keeps the best zero-shot topic if the cosine similarity with the topic meets `zeroshot_min_similarity`. Otherwise, the calculated representation is used.

Fixes #1967

MaartenGr commented on bertopic/_bertopic.py:
```diff
@@ -256,22 +256,41 @@ def __init__(self,
         self.topic_representations_ = None
         self.topic_embeddings_ = None
-        self.topic_labels_ = None
+        self.topic_id_to_zeroshot_topic_idx = {}
```
I'm using the `_` at the end of the variable to distinguish between parameters/public attributes/private attributes. Could you update this so it is inline with the structure of the public attributes? Unless you believe this variable should not be accessed by the user at all, then we can convert it to a private attribute.

ianrandman replied:
You're right. Forgot about this. Made it private.

MaartenGr commented on bertopic/_bertopic.py:
```diff
@@ -439,14 +466,23 @@ def fit_transform(self,
             # Save the top 3 most representative documents per topic
             self._save_representative_docs(documents)
 
+        # In the case of zero-shot topics, probability will come from cosine similarity,
+        # and the HDBSCAN model will be removed
+        if self._is_zeroshot() and len(assigned_documents) > 0:
+            self.hdbscan_model = BaseCluster()
+            sim_matrix = cosine_similarity(embeddings, np.array(self.topic_embeddings_))
+
+            if self.calculate_probabilities:
+                probabilities = sim_matrix
+            else:
+                # Use `topics_before_reduction` because `self.topics_` may have already been updated from
+                # reducing topics, and the original probabilities are needed for `self._map_probabilities()`
+                probabilities = sim_matrix[np.arange(len(documents)), topics_before_reduction]
```
Could this be simplified by moving it to the end of the function and running something like:

```python
if self._is_zeroshot() and len(assigned_documents) > 0:
            self.hdbscan_model = BaseCluster()
            _, probabilities = self.transform(assigned_documents, embeddings)
```

That way, it would actually be a more proper `.fit` + `.transform` combo as you typically see in scikit-learn and reduces the lines of code a bit.

ianrandman replied:
This goes back to our discussion on using cosine similarities for probabilities after zero-shot topic modeling.

> They are indeed not perfectly aligned with one another. With flat probabilities/similarities, this is not an issue as we can simply give back the similarity of the assigned topic. With topic-doc probabilities/similarities, I'm not sure how to solve that issue without attempting to create a separate model (i.e., logistic regression) that is trained (and perhaps even overfitted) on the embeddings (X) and assigned topic (y) to extract hopefully more accurate probabilities.

If using `.transform`, the assignments, and the corresponding probabilities will not be aligned with the current state of assignments, because the current state came from a mix of zero-shot and clustering, and `self.topic_embeddings_` was not used at all when making these assignments.

We agreed this misalignment is fine. We will return the cosine similarities for the current assignments, even if calculating similarities using `self.topic_embeddings_` would show different assignments. I am not sure it is possible to reuse `.transform` for this purpose. I'm open to suggestions, but having that bit of duplicated code seems like the simplest solution.

On a related note, I do not think my solution accounts for outliers.
I have
```
probabilities = sim_matrix[
                    np.arange(len(documents)), topics_before_reduction
                ]
```
but I think it should be
```
probabilities = sim_matrix[
                    np.arange(len(documents)), np.array(topics_before_reduction) + self._outliers
                ]
```
Does this make sense to you?

MaartenGr replied:
> From a user perspective, I would expect the assignments from .fit_transform to be identical to the assignments from .fit followed by .transform.

Agreed, that wasn't the case previously (also due to some specific behavior of HDBSCAN) but that would definitely be preferred. 

> I'm open to suggestions, but having that bit of duplicated code seems like the simplest solution. 

Ah right, I totally missed that! Yep, let's keep it as is. 

> On a related note, I do not think my solution accounts for outliers.

Yeah, that should indeed be `+ self._outliers` I believe since the lowest potential value is -1 there which does nasty things when indexing.

MaartenGr commented on bertopic/_bertopic.py:
```diff
@@ -3565,168 +3593,62 @@ def _zeroshot_topic_modeling(self, documents: pd.DataFrame, embeddings: np.ndarr
+                                 embeddings: np.ndarray,
                                  assigned_documents: pd.DataFrame,
-                                 embeddings: np.ndarray) -> Union[Tuple[np.ndarray, np.ndarray], np.ndarray]:
-        """ Combine the zero-shot topics with the clustered topics
-
-        There are three cases considered:
-        * Only zero-shot topics were found which will only return the zero-shot topic model
-        * Only clustered topics were found which will only return the clustered topic model
-        * Both zero-shot and clustered topics were found which will return a merged model
-          * This merged model is created using the `merge_models` function which will ignore
-            the underlying UMAP and HDBSCAN models
-
-        Arguments:
-            documents: Dataframe with documents and their corresponding IDs
-            assigned_documents: Dataframe with documents and their corresponding IDs
-                                that were assigned to a zero-shot topic
-            embeddings: The document embeddings
```
I would like to keep the docstrings here as it helps me (and others) develop. We could also move this to `._zeroshot_topic_modeling` and add some small docstrings here to indicate that this is only run when we have both zeroshot topics and non-zeroshot topics.

ianrandman replied:
I had removed the docstring when making big changes and forgotten to write a new one. Added.

MaartenGr reviewed (commented):
Thank you for your work on this! I left a couple of small comments. Other than that, can you run `ruff`? With a PR that was recently merged, we now use `ruff` for the formatting/linting.

ianrandman commented:
> Thank you for your work on this! I left a couple of small comments. Other than that, can you run `ruff`? With a PR that was recently merged, we now use `ruff` for the formatting/linting.

Thanks for pointing out the recent incorporation with `ruff`. I'll have to start using that in my own projects.

I have fixed my changes with `ruff`, fixed the merge confict, and resolved a couple of your comments. Please mark them resolved if my changes look good. There is only the remaining discussion about probabilities.

freddyheppell commented:
It looks like this is still failing because I think you only ran one of the two Ruff commands.

Ruff has `format`, which replaces e.g. Black and `check` which replaces e.g. flake8. If you run `ruff check --fix` it should autofix the remaining issues and let you know which need manual fixing. You can also run `make format` and `make lint` as shortcuts.

MaartenGr commented:
I believe the code check in python 3.8 failed because its not familiar with the `tuple[..., ....]` type hints. I believe replacing them with either `from typing import Tuple` should work or simply removing those type hints.

MaartenGr commented:
@ianrandman Awesome, everything passed and I think we addressed all the comments we had. Just to be 100% sure, shall I go ahead and merge this?

ianrandman commented:
> @ianrandman Awesome, everything passed and I think we addressed all the comments we had. Just to be 100% sure, shall I go ahead and merge this?

Yes, all good to merge if it looks good to you. Happy to be done with this :).

MaartenGr commented:
@ianrandman Awesome, thank you for taking the time the last couple of works to work on this. It is greatly appreciated and hopefully this will also make it easier for you to use BERTopic instead of your own fork. If there are any other changes you would like to see, please let me know!
